crates/bevy_animation/src/animation_curves.rs

```diff
@@ -0,0 +1,544 @@
+
+/// This type allows the conversion of a [curve] valued in the [property type] of an
+/// [`AnimatableProperty`] into an [`AnimationCurve`] which animates that property.
+///
+/// [curve]: Curve
+/// [property type]: AnimatableProperty::Property
+#[derive(Reflect, FromReflect)]
+#[reflect(from_reflect = false)]
+pub struct AnimatableCurve<P, C> {
+    curve: C,
+    #[reflect(ignore)]
+    _phantom: PhantomData<P>,
+}
+
+impl<P, C> AnimatableCurve<P, C>
+where
+    P: AnimatableProperty,
+    C: InnerAnimationCurve<P::Property>,
+{
+    /// Create an [`AnimatableCurve`] (and thus an [`AnimationCurve`]) from a curve
+    /// valued in an [animatable property].
+    ///
+    /// [animatable property]: AnimatableProperty::Property
+    pub fn from_curve(curve: C) -> Self {
+        Self {
+            curve,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<P, C> Clone for AnimatableCurve<P, C>
+where
+    C: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            curve: self.curve.clone(),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<P, C> Debug for AnimatableCurve<P, C>
+where
+    C: Debug,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("AnimatableCurve")
+            .field("curve", &self.curve)
+            .finish()
+    }
+}
+
+impl<P, C> AnimationCurve for AnimatableCurve<P, C>
+where
+    P: AnimatableProperty,
+    C: InnerAnimationCurve<P::Property>,
+{
+    fn clone_value(&self) -> Box<dyn AnimationCurve> {
+        Box::new(self.clone())
+    }
+
+    fn domain(&self) -> Interval {
+        self.curve.domain()
+    }
+
+    fn apply<'a>(
+        &self,
+        t: f32,
+        _transform: Option<Mut<'a, Transform>>,
+        mut entity: EntityMutExcept<'a, (Transform, AnimationPlayer, Handle<AnimationGraph>)>,
+        weight: f32,
+    ) -> Result<(), AnimationEvaluationError> {
+        let mut component = entity.get_mut::<P::Component>().ok_or_else(|| {
+            AnimationEvaluationError::ComponentNotPresent(TypeId::of::<P::Component>())
+        })?;
+        let property = P::get_mut(&mut component)
+            .ok_or_else(|| AnimationEvaluationError::PropertyNotPresent(TypeId::of::<P>()))?;
+        let value = self.curve.sample_clamped(t);
+        *property = <P::Property>::interpolate(property, &value, weight);
+        Ok(())
+    }
+}
+
+/// This type allows a [curve] valued in `Vec3` to become an [`AnimationCurve`] that animates
+/// the translation component of a transform.
+///
+/// [curve]: Curve
+#[derive(Debug, Clone, Reflect, FromReflect)]
+#[reflect(from_reflect = false)]
+pub struct TranslationCurve<C>(pub C);
+
+impl<C> AnimationCurve for TranslationCurve<C>
+where
+    C: InnerAnimationCurve<Vec3>,
+{
+    fn clone_value(&self) -> Box<dyn AnimationCurve> {
+        Box::new(self.clone())
+    }
+
+    fn domain(&self) -> Interval {
+        self.0.domain()
+    }
+
+    fn apply<'a>(
```

Comment:
I wonder, why is the function signature like this?
Can't Transform be lumped into the branch for any other component?
Is this for perf reasons?

Reply:
As far as I know it's for performance reasons that I don't fully grasp. :)

Reply:
Yes, it's for performance reasons. `EntityMut` has to do access checks on its components every time you access them, while passing `Transform` like this has no access checks.